### ISSUE TITLE
Log continuously models with wandb

### ISSUE THREAD

**Issue by borisdayma** [huggingface/transformers#10827]
# 🚀 Feature request

wandb integration currently logs last model (which can be the best by using `TrainingArguments.load_best_model_at_end`).

It would be great to allow continuous upload of model with appropriate aliases to versions.

Options would be:
    * `WANDB_LOG_MODEL = True` which just logs at the end as currently (not sure if we want to add scheduler and optimizer)
    * `WANDB_LOG_MODEL = 'all'` which logs continuously the model
    * `WANDB_LOG_MODEL = False` which does not log the model

## Motivation

Training can be very long and it would be so sad to lose a model :sob: 

## Your contribution

I can probably propose a PR but would love brainstorming on the ideal logic:
1. should we leverage `Trainer.save_model` (as currently) or `Trainer._save_checkpoint`
2. should we consider an artifact version as containing only the model & config or also containing optimizer and scheduler? Or should it actually be 2 separate artifacts?
3. if we leverage `on_save`, can we avoid the same current logic (fake trainer saving to a temporary directory that is then uploaded async) and just use an actual copy of what has been saved. We would just need the path or list of files that have been saved (should be straightforward)
4. If we log continuously the model, should we upload it only if it's improved (when `metric_for_best_model` is defined)? If that's the case, we'll need to be able to detect when that is the case. If that's not the case we'll still need to be able to know which one is the best.

**Comment by borisdayma** [comment]
I'm realizing now it would be so helpful with xlsr trainings (lost models due to crash after very long training).

Would you have any input or suggestions @sgugger on how I could try to implement it?

**Comment by sgugger** [comment]
Hi @borisdayma, I hadn't seen this issue. On our side we're more focused on how to continuously push the checkpoints to our own hub ;-).

That being said, we can definitely leverage the `on_save` event and just look for the last `checkpoint-xxx` folder, then push its content as artifact. If you have a tracking with `metric_for_best_model`, then you won't even have to look for the checkpoint, it will be in the state with `state.best_model_checkpoint`.
As for having one or separate checkpoints, I guess it really depends on what you think is best for WandB, you have more expertise than me here.

**Comment by borisdayma** [comment]
Thanks, it makes sense!
Actually I imagined you may probably have some interest in a similar logic for the model hub so I wanted to work on something that would be useful for everyone.

As a side note, models could also be stored on the model hub AND tracked by W&B (some people do it with ASW S3 for example). In this way, only checksums are actually stored and the files point back to the storage space so it could be complementary.

**Comment by sgugger** [comment]
To do the same on the hub, my idea was to leverage the versioning system and just push the saved checkpoint every save with a commit message like "checkpoint step xxx". Ideally inside a Callback to avoid adding more stuff to the main training loop.
I'll try to focus on this next week and see what we can easily do!

**Comment by borisdayma** [comment]
Nice, if possible it would be cool to allow a logger method to be called right after you push your checkpoint.

**Comment by borisdayma** [comment]
Still interested in working on it! Let me know when you have a hook for the model hub!

**Comment by sgugger** [comment]
We have started working on it (the Trainer gained a push_to_hub API!) the last step missing is the continuous checkpointing, will try to work on that last part soon!

**Comment by borisdayma** [comment]
Awesome, if you can somehow have a hook after pushing to the hub (with access to the url maybe) then we could link link them to the runs.

**Comment by sgugger** [comment]
The call to `Trainer.push_to_hub` returns the url of the commit to the model hub (it's not part of the train for now).

**Comment by borisdayma** [comment]
I really like where this is going!

Is the goal for `TrainingArguments.push_to_hub` to be eventually directly used by the `Trainer` or will it always be handled by the scripts?
Also would it be possible to save the return of `_push_to_hub` somewhere in the `Trainer` (that way that url could be used by wandb).
